### PR TITLE
Hashes for `Quad`

### PR DESCRIPTION
**Pull request by StevellM**
Complements #1100 

- Change of the comparison test for `AbstractLat` which brings a broken test as mentioned in #1112 
- Add `==` for `TorQuadModule` which are defined with by lattices which all are in the same ambient (quadratic) space.

### CODE REVIEW

File: src/QuadForm/Herm/Genus.jl [hunk 1]
```diff
@@ -1322,6 +1345,8 @@ function direct_sum(G1::HermGenus, G2::HermGenus)
   sig1 = G1.signatures
   sig2 = G2.signatures
   g3 = merge(+, sig1, sig2)
+  bd = union(support(2*maximal_order(base_field(E))), support(discriminant(maximal_order(E))))
+  filter!(g -> (prime(g) in bd) || (scales(g) != Int[0]), LGS)
```
Comment (thofma) [inline src/QuadForm/Herm/Genus.jl:1349]: Is the `scales(g) != Int[0]` for the block not being unimodular?

**Comment by thofma** [comment]
I think this all looks good, but we need some comments to remember what we did in a few weeks (a few days?).

Somewhere we need to write down which primes we throw away etc.

File: src/QuadForm/Herm/Genus.jl [hunk 2]
```diff
@@ -1018,6 +1020,21 @@ function ==(G1::HermLocalGenus, G2::HermLocalGenus)
...
+  u = Base.hash(base_field(g), u)
+  u = Base.hash(prime(g), u)
+  h = xor(hash(det(g)), reduce(xor, (hash(s[1:2]) for s in g.data)))
+  if is_ramified(g)
+    if !is_dyadic(g)
+      h = xor(h, reduce(xor, (hash(s[3]) for s in g.data if iseven(s[1]))))
+    else
+      h = xor(h, reduce(xor, (hash(g.data[i][1] == 2*g.norm_val[i]) for i = 1:length(g.data))))
+      h = xor(h, hash(g.ni))
+    end
```
Comment (simonbrandhorst) [inline src/QuadForm/Herm/Genus.jl:1033]: Please explain.

File: src/QuadForm/Quad/Genus.jl [hunk 3]
```diff
@@ -758,10 +763,22 @@ end
...
 
+function Base.hash(g::QuadLocalGenus, u::UInt)
+  u = Base.hash(base_field(g), u)
+  u = Base.hash(prime(g), u)
+  h = xor(hash(scales(g)), hash(ranks(g)))
+  if is_dyadic(g)
+    h = xor(h, hash(weights(g)))
+    h = xor(h, hash(witt_invariant(g)))
+    h = xor(h, hash(norms(g)))
+  end
```
Comment (simonbrandhorst) [inline src/QuadForm/Quad/Genus.jl:778]: Are the norms really invariants?

**Review by simonbrandhorst — COMMENTED** [review]
Please enforce the sparsity of the local genus symbols in the constructor.
Either by a test or more conveniently by filtering the symbol in the constructor method.

File: src/QuadForm/Herm/Genus.jl (hunk 1 above)
Comment (StevellM) [reply src/QuadForm/Herm/Genus.jl:1349]: Exactly, we keep the local symbol if the prime of bad, or if it is not unimodular. A local symbol is unimodular if and only if it consists of only one block of scale zero.

**Comment by StevellM** [comment]
I will add a bunch of comments regarding what we discard and the choices made for the hashes. I will also re-enable the broken test since the issue has be solved.

**Review by simonbrandhorst — APPROVED** [review]

**Pull request merged**